### DISCUSSION TITLE
[Thesis] Multiplayer

### DISCUSSION THREAD

**Discussion by AMatutat** [Programmiermethoden/Dungeon#1369] [Ideas]
Can we make some kind of MMORPG out of this?

**Comment by AMatutat** [comment]
Info: Hier wird aktuell im Rahmen eines Forschungsprojektes dran gearbeitet

**Comment by yogo1212** [comment]
# Paradigmen zur Abgleichung der Simulation

Ein Vergleich von grundlegenden Multiplayer-Paradigmen findet sich in [diesem Paper](https://hdms.bsz-bw.de/frontdoor/deliver/index/docId/7107/file/DeterministicLockstepInNetworkedGamesPaper.pdf)  (Tabelle auf Seite 3):

([ausführlichere Vorstellung der Paradigmen](https://gafferongames.com/categories/networked-physics/))

Eine Gegenüberstellung vor dem Hintergrund der Verwendung im Dungeon:

## Snapshots

- Abgleich der Simulationen anhand vollständiger Zustandsbeschreibungen
- Verwendung zum Speichern/Laden oder Hot-Join möglich
- bei ausschließlicher Verwendung:
  - quasi keine Probleme mit Paket-Loss
  - geringe Latenz
  - Übertragung mehrmals in der Sekunde
  - dafür muss der Zustand der Simulation sehr effizient über das Netzwerk übertragbar sein (Bandbreite + CPU)
  - laufende Optimierung der Zustandsbeschreibung wahrscheinlich schwierig im Dungeon
  - ein Server verarbeitet die Eingaben der Teilnehmer 
  - prinzipiell ungeeignet für P2P

Typische Einsatzszenarien für snapshot interpolation: wenig Einheiten in der Simulation, hohe Anforderungen an Latenz. Zum Beispiel Shooter oder MOBAs.

## Deterministic Lockstep

- Kommunikation geänderter Eingabeparameter für die Simulation (z.B. Nutzereingaben)
- dadurch meistens geringe Bandbreitenanforderung
- "lockstep" bezeichnet schrittweise synchronisiertes Voranschreiten der Simulation
- bei ausschließlicher Verwendung:
  - die Parameter müssen zuverlässig vor jedem Schritt der Simulation ausgetauscht werden
  - -> höhere Latenz
  - verträglich mit P2P

Typische Einsatzszenieren: Umfangreiche Simulation (mehr Einheiten), geringere Anforderungen an Latenz. Zum Beispiel Strategie- oder Simulationsspiele.

## Kombination aus Beidem

- Mehraufwand bei der Implementierung
  - Aspekte von beiden Paradigmen müssen implementiert werden
  - zusätzlicher Aufwand z.B. für Übertragungen von Zuständen von RNGs
- geringer Optimierungsbedarf in beiden Systemen
- gelockerte Anforderungen an die Snapshots
- gelockerte Anforderungen an die deterministische Simulation
  - regelmäßige Snapshots bereinigen Diskrepanzen in eventuell nicht ganz identischen Simulationen
- nicht mehr so gut verträglich mit P2P

# Netzwerkarchitektur

## Netzwerktopologie

Unabhängig von der Topologie ist der Verbindungsaufbau.

### P2P

- die Topologie entsteht dynamisch zur Laufzeit
  - der Netzwerkgraph hat vielleicht nicht alle wünschenwerten Eigenschaften
  - Implementierungsaufwand für die Verwaltung des Netzwerks zur Laufzeit
    - Zahl der möglichen Verbindungen wächst quadratisch mit der Zahl der Teilnehmer
    - Propagation neuer Teilnehmer, welche neuen Verbindungen sollen aufgebaut werden?
    - Verbindungsabbruch
    - ausbleibende Verbindungen
- keine implizite Authorität im Netzwerk
  - vielleicht dennoch wünschenswert (vlt. für Snapshots)
- Bottleneck der "naiven" Implementierung ist der Upstream jedes Ereignis-generierenden Teilnehmers
  - Größe des Netzwerk direkter Faktor für den Ausgangstraffic

### Client-Server

- Topologie ist immer ein Stern (Baum mit maximaler Höhe 1)
  - einfacher zu verwalten: Verbindung ist da oder nicht.
- die "Server"-Rolle kann dynamisch verschoben werden
  - bspw. wenn der aktuelle Host wegfällt

## Verbindungsaufbau

Es wird ein Einstiegspunkt benötigt. Meistens ist das ein öffentlich erreichbarer Broker (z.B. Lobby-/Master-Server). Bei der Verbindung von Teilnehmern an üblichen Internetanschlüssen ist zusätzlich ein Mechanismus zur Überwindung von NAT in Gateways und Firewalls nötig.
Konzeptionell ist die Kommunikation über Dritte sowohl für Client-Server als auch P2P nötig. Behelfsmäßig kann man auch IP-Adressen/Hostnamen "von Hand" (Email/Chat/Telefon) austauschen.

Für die ersten Versionen sind manuelle geöffnete Ports und manuelle Verbindungsaufbauten wahrscheinlich ein guter Kompromis. Für den langfristigen produktiven Einsatz wird eine integrierte Lösung vermutlich nötig sein. Auch dafür sind Kompromisse denkbar, beispielsweise der Einsatz von öffentlichen MQTT-Brokern zur Findung und die Nutzung von öffentlichen STUN-Servern für den Verbindungsaufbau.

**Comment by yogo1212** [comment]
@cagix Ist das hier der richtige Ort? Ich habe im Sinne von `- Marketing` den Text stark umgeformt.

**Reply by cagix** [reply]
@yogo1212 passt perfekt. danke dir 🙏

**Comment by AMatutat** [comment]
haben wir ja jetzt
